fms/utils/generation.py

```diff
@@ -20,9 +20,73 @@ def _make_cache_contiguous(past_key_value_states):
     return n_kv_s
 
 
+def left_pad(input_ids: torch.Tensor, max_length: int, pad_id: int = 0) -> torch.Tensor:
+    """
+    left pad an input_ids tensor
+
+    Parameters
+    ----------
+    input_ids: torch.Tensor
+        input ids corresponding to a single sequence in a batch
+    max_length: int
+        the max length to pad to
+    pad_id: int
+        the token to set as a pad in the resulting tensor
+
+    Returns
+    -------
+    torch.Tensor
+        a left padded tensor
+    """
+    pads_tensor = torch.tensor(
+        [pad_id] * (max_length - input_ids.size(0)),
+        device=input_ids.device,
+        dtype=torch.long,
+    )
+    return torch.cat((pads_tensor, input_ids))
+
+
+def __create_attention_mask(
+    input_ids: torch.Tensor, use_cache: bool, is_prompt: bool, pad_id: int = -1
+) -> Optional[torch.Tensor]:
+    """
+    Optionally create an attention mask if one is required
+
+    Parameters
+    ----------
+    input_ids: torch.Tensor
+        the input ids to the model's forward method
+    use_cache: bool
+        if True caching is being used, otherwise caching is not being used
+    is_prompt: bool
+        if True, input_ids correspond to the prompt, otherwise input_ids contain generated tokens
+    pad_id: int
+        the pad id used when padding. If the pad_id is -1, no padding is being used, otherwise assume some padding
+
+    Returns
+    -------
+    Optional[torch.Tensor]
+        if is_prompt, no use_cache, batch required padding => a bool Tensor is created for the mask
+        otherwise None is returns and the mask will be handled internally as part of sdpa
+    """
+    # compute the attention mask
+    # always need a mask if this is the prompt or if not using a cache
+    if is_prompt or not use_cache:
```

Comment:
this condition is the only place either of these booleans are used, so maybe should just be a single boolean? The descriptions in the comments also don't really clarify their meaning/purpose, so maybe a single combined bool could be clearer about the intent here.